ci/ray_ci/test_upgrade_version.py

```diff
@@ -0,0 +1,99 @@
+from unittest import mock
+import sys
+import pytest
+import tempfile
+
+from ci.ray_ci.upgrade_version import (
+    list_java_files,
+    get_current_version,
+    upgrade_file_version,
+)
+
+
+@mock.patch("os.walk")
+def test_list_java_files(mock_os_walk):
+    mock_os_walk.return_value = [
+        (
+            "root/dir",
+            ["subdir_1", "subdir_2", "subdir_3"],
+            ["pom_template.xml", "not_pom.xml"],
+        ),
+        ("root/dir/subdir_1", [], ["pom.xml", "not_pom.xml", "not_pom_template.xml"]),
+        ("root/dir/subdir_2", [], ["pom_template.xml", "pom.xml", "not_pom.py"]),
+        (
+            "root/dir/subdir_3",
+            [],
+            ["pom.xml", "pom_template.xml", "not_pom_template.py"],
+        ),
+    ]
+    assert list_java_files() == {
+        "root/dir/pom_template.xml",
+        "root/dir/subdir_1/pom.xml",
+        "root/dir/subdir_2/pom_template.xml",
+        "root/dir/subdir_2/pom.xml",
+        "root/dir/subdir_3/pom.xml",
+        "root/dir/subdir_3/pom_template.xml",
+    }
+
+
+@mock.patch("subprocess.run")
```

Comment:
rather than mocking `subprocess.run`, define a small wrapper around the call (should be `check_output` now), and mock that.

better if you can send in an optional function pointer to the function under test.

Reply:
I'm not sure if it's allowed to mock nested function (if that's what you mean by defining a wrapper around the call). 
I have separated that call into another function and mock that instead. Please correct me if it's not the right way.